reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java

```diff
@@ -0,0 +1,829 @@
+		return acquired;
+	}
+
+	@Override
+	public int allocatedSize() {
+		return acquired;
+	}
+
+	@Override
+	public PoolConfig<Connection> config() {
+		return poolConfig;
+	}
+
+	@Override
+	public Mono<Void> disposeLater() {
+		return Mono.defer(() -> {
+			recordInteractionTimestamp();
+
+			@SuppressWarnings("unchecked")
+			ConcurrentLinkedDeque<Borrower> q = PENDING.getAndSet(this, TERMINATED);
+			if (q != TERMINATED) {
+				Borrower p;
+				while ((p = q.pollFirst()) != null) {
+					p.fail(new PoolShutdownException());
+				}
+
+				// the last stream on that connection will release the connection to the parent pool
+				// the structure should not contain connections with 0 streams as the last stream on that connection
+				// always removes the connection from this pool
+				CONNECTIONS.getAndSet(this, null);
+			}
+			return Mono.empty();
+		});
+	}
+
+	@Override
+	public int getMaxAllocatedSize() {
+		return Integer.MAX_VALUE;
+	}
+
+	@Override
+	public int getMaxPendingAcquireSize() {
+		return poolConfig.maxPending() < 0 ? Integer.MAX_VALUE : poolConfig.maxPending();
+	}
+
+	@Override
+	public int idleSize() {
+		return 0;
+	}
+
+	@Override
+	public boolean isDisposed() {
+		return PENDING.get(this) == TERMINATED || CONNECTIONS.get(this) == null;
+	}
+
+	@Override
+	public boolean isInactiveForMoreThan(Duration duration) {
+		return pendingAcquireSize() == 0 && allocatedSize() == 0
+				&& secondsSinceLastInteraction() >= duration.getSeconds();
+	}
+
+	@Override
+	public PoolMetrics metrics() {
+		return this;
+	}
+
+	@Override
+	public int pendingAcquireSize() {
+		return PENDING.get(this).size();
+	}
+
+	@Override
+	public long secondsSinceLastInteraction() {
+		long sinceMs = clock.millis() - lastInteractionTimestamp;
+		return sinceMs / 1000;
+	}
+
+	@Override
+	public Mono<Integer> warmup() {
+		return Mono.just(0);
+	}
+
+	void cancelAcquire(Borrower borrower) {
+		if (!isDisposed()) {
+			ConcurrentLinkedDeque<Borrower> q = pending;
+			q.remove(borrower);
+		}
+	}
+
+	Mono<Void> destroyPoolable(Http2PooledRef ref) {
+		Mono<Void> userProvidedDestroy;
+		try {
+			userProvidedDestroy = Mono.from(destroyHandler().apply(this, ref));
+		}
+		catch (Throwable destroyFunctionError) {
+			userProvidedDestroy = Mono.error(destroyFunctionError);
+		}
+		return userProvidedDestroy;
+	}
+
+	BiFunction<Http2Pool, Http2PooledRef, Publisher<Void>> destroyHandler() {
+		return (pool, pooledRef) -> {
+			pooledRef.slot.decrementConcurrency();
+			Connection connection = pooledRef.poolable();
+			Http2FrameCodec frameCodec = connection.channel().pipeline().get(Http2FrameCodec.class);
+			if (frameCodec != null) {
+				if (pooledRef.slot.concurrency() == 0 && pooledRef.slot.invalidate()) {
+					releaseConnection(connection);
+				}
+			}
+			else {
+				// no need to guard here with invalidate()
+				// #findConnection will never give a connection that does not have Http2FrameCodec
+				poolConfig.allocationStrategy().returnPermits(1);
```

Comment:
I need to have a harder look at this part, but I think I'm getting my head around it. I wonder if `returnPermit` outside of the one done in `invalidate` really makes sense. Would appreciate if you can detail the logic a bit more, though.

Reply:
This is for handling the cases where the negotiated protocol is HTTP/1.1

There is a test for this use case
reactor.netty.http.client.Http2PoolTest#nonHttp2ConnectionEmittedOnce

invalidate() makes sense only when we are in HTTP/2 use case.

Reply:
ok, so it boils downs to the fact that we can still end up serving "slots" to HTTP/1.1 clients, in which case we never increment the slot usage nor do we want to keep it around, correct?

but since  `invalidate` does returnPermits + removeSlot under the cover, wouldn't it simplify the code to always call `invalidate` here ?

that is, provided `concurrency` is kept at `1` for the HTTP/1.1 case...

something like:

```java
Http2FrameCodec frameCodec = connection.channel().pipeline().get(Http2FrameCodec.class);
if (ref.slot.decrementConcurrencyAndGet() == 0 //more on this below
  && ref.slot.invalidate()) {
	if (frameCodec != null) {
		//case serving HTTP/2 client
		releaseConnection(connection);
	} //else serving HTTP/1.1 client, no need to releaseConnection
}
```

I'm a bit worried that calls to `decrementConcurrency` and `concurrency()` are split, as this negates the benefits of the underlying atomic: you increment and check the value separately. I think it would be preferable to expose more atomic operations, like prototyped in the snippet above (see my other comment in `Slot` too).